pkg/providers/gcp/cluster_network_provider.go

```diff
@@ -1,24 +1,394 @@
+	projectID, err := resources.GetGCPProject(ctx, n.Client)
 	if err != nil {
-		return "Could not create network", err
+		return errorUtil.Wrap(err, "error getting project name")
 	}
-	return croType.StatusEmpty, nil
+	// build ip address range name
+	ipRange, err := resources.BuildInfraName(ctx, n.Client, defaultIpRangePostfix, defaultGcpIdentifierLength)
+	if err != nil {
+		return errorUtil.Wrap(err, "failed to build ip address range infra name")
+	}
+	// get ip address range
+	address, err := n.getAddressRange(ctx, projectID, ipRange)
+	if err != nil {
+		return errorUtil.Wrap(err, "failed to retrieve ip address range")
+	}
+	if address != nil {
+		clusterVpc, err := getClusterVpc(ctx, n.Client, n.NetworkApi, n.Logger)
+		if err != nil {
+			return errorUtil.Wrap(err, "failed to get cluster vpc")
+		}
+		// get service connection
+		service, err := n.getServiceConnection(clusterVpc, projectID)
+		if err != nil {
+			return err
+		}
+		if service != nil && service.ReservedPeeringRanges[0] == address.GetName() {
+			return errors.New("failed to delete ip address range, service connection still present")
+		}
+		// delete ip address range
+		err = n.deleteAddressRange(ctx, projectID, ipRange)
+		if err != nil {
+			return errorUtil.Wrap(err, "failed to delete ip address range")
+		}
+	}
+	return nil
+}
+
+func (n *NetworkProvider) ComponentsExist(ctx context.Context) (bool, error) {
+	clusterVpc, err := getClusterVpc(ctx, n.Client, n.NetworkApi, n.Logger)
+	if err != nil {
+		return false, errorUtil.Wrap(err, "failed to get cluster vpc")
+	}
+	// get project ID
+	projectID, err := resources.GetGCPProject(ctx, n.Client)
+	if err != nil {
+		return false, errorUtil.Wrap(err, "error getting project name")
+	}
+	// build ip address range name
+	ipRange, err := resources.BuildInfraName(ctx, n.Client, defaultIpRangePostfix, defaultGcpIdentifierLength)
+	if err != nil {
+		return false, errorUtil.Wrap(err, "failed to build ip address range infra name")
+	}
+	// get ip address range
+	address, err := n.getAddressRange(ctx, projectID, ipRange)
+	if err != nil {
+		return false, errorUtil.Wrap(err, "failed to retrieve ip address range")
+	}
+	if address != nil {
+		n.Logger.Infof("ip address range %s deletion in progress", address.GetName())
+		return true, nil
+	}
+	// get service connection
+	service, err := n.getServiceConnection(clusterVpc, projectID)
+	if err != nil {
+		return false, err
+	}
+	if service != nil {
+		n.Logger.Infof("service connection %s deletion in progress", service.Service)
+		return true, nil
+	}
+	return false, nil
+}
+
+func (n *NetworkProvider) getServiceConnection(clusterVpc *computepb.Network, projectID string) (*servicenetworking.Connection, error) {
+	resp, err := n.ServicesApi.ConnectionsList(clusterVpc, projectID, fmt.Sprintf("services/%s", defaultServiceConnectionURI))
+	if err != nil {
+		return nil, err
+	}
+	if len(resp.Connections) == 0 {
+		return nil, nil
+	}
+	return resp.Connections[0], nil
+}
```

Comment:
Is there a possibility of multiple service connections being returned by `ConnectionsList()`? If that happens, how would we determine we're using the right one?

Reply:
We filter the list based on the specific service connection name, and you can only have one active connection to the service so it should not be possible to return multiple. This is a workaround for the lack of a direct GET method